python/ray/ml/trainer.py

```diff
@@ -54,6 +142,69 @@ def __init__(
 
         raise NotImplementedError
 
+    @DeveloperAPI
+    def setup(self) -> None:
+        """Perform initial setup on the Trainer.
+
+        Note: this method is run on a remote process.
+
+        This method will not be called on the driver, so any expensive setup
+        operations should be placed here and not in ``__init__``.
+
+        This method is called prior to ``_preprocess_datasets`` and
+        ``training_loop``.
+        """
+        raise NotImplementedError
+
+    def _preprocess_datasets(self) -> None:
```

Comment:
Can we make this DeveloperAPI as well? It's odd that this one is private.

Reply:
The way I'm denoting is like so:

PublicAPI- APIs for end user
DeveloperAPI- API for developers to create new Trainers. So setup and training_loop can/must be overridden.

_preprocess_datasets does not fall into either of those. It should not be called by end users, and should not be overridden by developers.

Reply:
Why shouldn't it be overriden? DataParallelTrainer might create ingest pipelines and things like that. How about we just keep it like that for consistency / generalizability?

Reply:
Got it, makes sense- updated!